src/platforms/common/stm32/swo_manchester_dma.c

```diff
@@ -0,0 +1,570 @@
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
+ * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+ * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+/*
+ * This is an alternate high-performance implementation of Manchester-encoded
+ * SWO trace output for the Black Magic Debug Probe.
+ *
+ * This is a rough outline of the capture process:
+ *
+ * 1. all edge times of the signal are captured using a timer
+ * 2. DMA is used to record the timings into a circular buffer
+ * 3. the buffer is periodically processed in batches, transformig the edge
+ *    stream into a byte stream for sending in another circular buffer, 
+ *    resulting in effective processing time per sample on the order of
+ *    several clock cycles
+ * 4. the output buffer is processed in a lower-priority ISR as time permits
+ *
+ * The decoding is reasonably reliable for SWO frequencies from 100 kHZ up to
+ * ~3 MHz and is mostly resilient against noise on the SWO input
+ * 
+ */
+
+#include "general.h"
+#include "platform.h"
+#include "usb.h"
+#include "swo.h"
+#include "swo_internal.h"
+
+#include <libopencm3/cm3/nvic.h>
+#include <libopencm3/stm32/timer.h>
+#include <libopencm3/stm32/rcc.h>
+#include <libopencm3/stm32/dma.h>
+#include <libopencm3/stm32/st_usbfs.h>
+
+// Number of 16-bit samples captured for processing, the processing
+// is triggered when half of the buffer is full, must be a power of two
+#ifndef SWO_DMA_EDGE_SAMPLES
+#define SWO_DMA_EDGE_SAMPLES	512
+#endif
+
+#define SWO_DMA_EDGE_MASK		((SWO_DMA_EDGE_SAMPLES)-1U)
+
+#define SWO_BUFFER_MASK			((SWO_BUFFER_SIZE)-1U)
+
+// Fixed length of a pulse that resets the decoder
+// 72 MHz / 4096 = ~17.6 kHZ longest acceptable pulse
+// limiting the lowest processable frequency to about 40 kHZ,
+// but there is little reason to do that
+#ifndef SWO_MAX_PULSE
+#define SWO_MAX_PULSE 4096
+#endif
+
+// Include advanced recovery logic
+// This makes sense only for very low frequencies when there is never
+// a proper break in the pulse sequence
+//#define SWO_ADVANCED_RECOVERY	1
+
+#define FORCE_INLINE	inline __attribute__((always_inline))
+
+// edge time buffer
+static uint16_t swo_dma[SWO_DMA_EDGE_SAMPLES];
+
+// helper to get the current write head of the DMA
+static FORCE_INLINE uint16_t swo_dma_wx()
+{
+	return SWO_DMA_EDGE_SAMPLES - DMA_CNDTR(SWO_DMA_BUS, SWO_DMA_EDGE_CHAN);
+}
+
+static FORCE_INLINE void swo_buffer_write(uint8_t b)
+{
+	uint16_t wx = swo_buffer_write_index;
+	swo_buffer[wx++] = b;
+	swo_buffer_write_index = wx & SWO_BUFFER_MASK;
+}
+
+static FORCE_INLINE void swo_buffer_push(void)
+{
+	// just trigger the IRQ and let it check if it makes sense to do more -
+	// it will be tail-chained after the main ISR anyway so an additional
+	// check now would be a waste
+	NVIC_STIR = SWO_DMA_SW_IRQ;
+}
+
+// decoder state
+static struct state {
```

Comment:
Please define this at the top of file below the `#define`'s - same for the enum which follows.